highfidelity/hifi
# CR-1 21114 -- ScriptCache/ScriptEngine cleanup, loader states, accessible load errors

This PR contains several changes from the `require/module` work, separated out this way to split things up a little for code review and testing.

* Add new entity script statuses (PENDING, LOADING and UNLOADED).
* Provide specific network errors for failed scripts.
* Resolve FIXMEs related to script loading/caching; cleanup dead code.
* Refactor support for JavaScript syntax checking and exception reporting.
* Update the editor's "Server Script URL" panel to reflect the new statuses.

With @huffman's help I was able to get that last part working, which now doubles as a way to test the overall changes from the UI.

A cool thing for the user is that augmented errors show up in the tablet/editor panel for server scripts -- like when a syntax error, the problematic line number; or if a script fails to load, the underlying network reason.

### Testing:

* Launch the PR build of Interface + Sandbox.
* Perform a general smoke test to do with loading Client scripts
 *(app behavior should not have changed -- but Client scripts are now fetched using a different method so just want to be sure)*
* Rez a few Entities to work with.
* Try attaching a representative sample of script URLs (in the `Script URL` and `Server Script URL` fields, but not necessarily same time).
* For Server scripts -- detailed error messages should come back in the textarea:
![image](https://cloud.githubusercontent.com/assets/94753/22858917/3ec8cbfa-f09a-11e6-8a66-c3f93560335f.png)
* For regular entity scripts -- similar error messages should get reported to the debug log:
`02/11 20:41:08] [DEBUG] [hifi.scriptengine] "Bad syntax ([SyntaxError] Illegal character in https://httpbin.org/image/png:1(1))" "https://httpbin.org/image/png"`

test urls:
* https://httpbin.org/delay/10 => ~10 seconds of *Loading* => `Could not find constructor`
* http://httpbin.org/404 => `Failed to load script (Not Found)`
* [ftp://0.0.0.0:0000](ftp://0.0.0.0:0000) => `Failed to load script (AccessDenied)`
* [http:////:ptth](http:////:ptth) => `Failed to load script (ServerUnavailable)`
* https://httpbin.org/image/png => `...[SyntaxError] Illegal character in https://httpbin.org/image/png:1(1)`
* [/scripts/developer/tests/basicEntityTest/myEntityScript.js](https://raw.githubusercontent.com/highfidelity/hifi/master/scripts/developer/tests/basicEntityTest/myEntityScript.js) => *Running* (no error)
*Clear-out a populated Server Script URL* => *Unloaded* or *Not Running* (no error)
